pilco/models/mgpr.py

```diff
@@ -1,68 +1,85 @@
+            else:
+                self.models[i].data = (data[0], data[1][:, i:i+1])
 
     def optimize(self, restarts=1):
         if len(self.optimizers) == 0:  # This is the first call to optimize();
             for model in self.models:
                 # Create an gpflow.train.ScipyOptimizer object for every model embedded in mgpr
-                optimizer = gpflow.train.ScipyOptimizer(method='L-BFGS-B')
-                optimizer.minimize(model)
+                optimizer = gpflow.optimizers.Scipy()
+                optimizer.minimize(model.training_loss, model.trainable_variables)
                 self.optimizers.append(optimizer)
-                restarts -= 1
+        else:
+            for model, optimizer in zip(self.models, self.optimizers):
+                #session = optimizer._model.enquire_session(None)
+                optimizer.minimize(model.training_loss, model.trainable_variables)
 
         for model, optimizer in zip(self.models, self.optimizers):
-            session = optimizer._model.enquire_session(None)
-            best_parameters = model.read_values(session=session)
-            best_likelihood = model.compute_log_likelihood()
+            # session = optimizer._model.enquire_session(None)
+            best_params = {
+                "lengthscales" : model.kernel.lengthscales.value(),
+                "k_variance" : model.kernel.variance.value(),
+                "l_variance" : model.likelihood.variance.value()}
+            #best_likelihood = model.log_marginal_likelihood()
+            best_loss = model.training_loss()
             for restart in range(restarts):
                 randomize(model)
-                optimizer._optimizer.minimize(session=session,
-                            feed_dict=optimizer._gen_feed_dict(optimizer._model, None),
-                            step_callback=None)
-                likelihood = model.compute_log_likelihood()
-                if likelihood > best_likelihood:
-                    best_parameters = model.read_values(session=session)
-                    best_likelihood = likelihood
-            model.assign(best_parameters)
+                optimizer.minimize(model.training_loss, model.trainable_variables)
+                #likelihood = model.log_marginal_likelihood()
+                loss = model.training_loss()
+                if loss < best_loss:
+                    best_params["k_lengthscales"] = model.kernel.lengthscales.value()
+                    best_params["k_variance"] = model.kernel.variance.value()
+                    best_params["l_variance"] = model.likelihood.variance.value()
+                    #best_likelihood = likelihood
+                    best_loss = model.training_loss()
+            model.kernel.lengthscales.assign(best_params["lengthscales"])
+            model.kernel.variance.assign(best_params["k_variance"])
+            model.likelihood.variance.assign(best_params["l_variance"])
```

Comment:
There is no way anymore to pull all the parameters with a single command, as was done with `model.read_values(session=session)` before?

Reply:
So I thought, but I later found out that we indeed can. There is a gpflow.utilities.multiple_assign functionality that I missed when writing this

Reply:
Is this going to simplify the code? If so is it possible to switch to that and remove the comments?